test/basic.test.js

```diff
@@ -23,43 +23,214 @@ test('Connect-Subscribe-Publish-Disconnect 300 clients using WS and MQTT/MQTTS p
+      t.pass('Subscription has been restored')
+    }
+  })
+
+  await publisher.publish('my/topic', 'I\'m alive', { qos: 1 })
+
+  await helper.delay(200)
+  await publisher.end()
+})
+
+test('Client receives retained messages on connect', async function (t) {
+  t.plan(10)
+  t.tearDown(helper.closeBroker)
+
+  await helper.startBroker()
+
+  var publisher = await helper.startClient()
+
+  const options = { qos: 1, retain: true }
+
   for (let i = 0; i < 10; i++) {
-    subscribers.push(helper.startClient())
+    await publisher.publish('test/retained/' + i, i.toString(), options)
   }
 
-  subscribers = await Promise.all(subscribers)
+  await publisher.end()
+
+  publisher = await helper.startClient()
+
+  await publisher.subscribe('test/retained/#')
+
+  publisher.on('message', function (topic, message) {
+    t.pass('Retained message received')
+  })
+
+  await helper.delay(200)
+  await publisher.end()
+})
+
+test('Will message', async function (t) {
+  t.plan(1)
+  t.tearDown(helper.closeBroker)
+
+  await helper.startBroker()
+
+  var client = await helper.startClient('mqtt', {
+    will: {
+      topic: 'my/will',
+      payload: 'I\'m dead',
+      qos: 1,
+      retain: false
+    },
+    reconnectPeriod: 100
+  })
+
+  var client2 = await helper.startClient()
 
-  function onMessage (topic, message) {
-    if (topic === msg.topic) {
-      t.pass('message received')
+  await client2.subscribe('my/will', { qos: 1 })
+
+  client2.on('message', function (topic, message) {
+    if (topic === 'my/will') {
+      t.pass('Will received')
     }
+  })
+
+  client.stream.destroy()
+
+  await helper.delay(200)
+  await client.end()
+  await client2.end()
+})
+
+test('Wildecard subscriptions', async function (t) {
+  t.tearDown(helper.closeBroker)
+
+  await helper.startBroker()
+
+  const options = {
+    qos: 0,
+    retain: false
   }
 
-  for (let i = 0; i < subscribers.length; i++) {
-    const client = subscribers[i]
-    await client.subscribe(msg.topic)
-    client.once('message', onMessage)
+  var subscriptions = {
+    'a/#': {
+      a: true,
+      'a/b': true,
+      'a/b/c': true,
+      'b/a/c': false
+    },
+    'a/+/+': {
+      'a/b/c': true,
+      'a/a/c': true,
+      'a/b/c/d': false,
+      'b/c/d': false
+    }
   }
 
   var publisher = await helper.startClient()
+  var subscriber = await helper.startClient()
+
+  function testReceive (sub, pub, result) {
+    return new Promise((resolve) => {
+      var timeout = null
+
+      subscriber.once('message', function (topic, message) {
+        if (timeout) {
+          clearTimeout(timeout)
+        }
+
+        if (result && topic === pub) {
+          resolve(true)
+        } else {
+          resolve(false)
+        }
+      })
+
+      subscriber.subscribe(sub, options)
+        .then(() => publisher.publish(pub, 'Test wildecards', options))
+        .catch((e) => resolve(false))
+
+      if (!result) {
+        timeout = setTimeout(resolve(true), 200)
```

Comment:
Is this needed?

Reply:
When the message should NOT be received, I set a timeout, is there other way?

Reply:
Done in latest commit